File: src/com/machfour/macros/core/Food.java
package com.machfour.macros.core;

import com.machfour.macros.data.Column;
import com.machfour.macros.data.ColumnData;
import com.machfour.macros.data.Table;
import com.machfour.macros.data.Tables;
import com.sun.istack.internal.NotNull;
import com.sun.istack.internal.Nullable;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import static com.machfour.macros.data.Columns.FoodCol.*;

public class Food extends MacrosEntity<Food> {

    public static final List<Column<Food, String>> DESCRIPTION_COLUMNS = Arrays.asList(
        BRAND
        , COMMERCIAL_NAME
        , VARIETY_PREFIX_1
        , VARIETY_PREFIX_2
        , NAME
        , VARIETY_SUFFIX
        , NOTES
        , INDEX_NAME
    );

    private final List<Serving> servings;

    private NutritionData nutritionData;
    private FoodType foodType;

    public Food(ColumnData<Food> dataMap, boolean isFromDb) {
        super(dataMap, isFromDb);
        servings = new ArrayList<>();
        foodType = null;
        nutritionData = null;

    }

    public FoodType getFoodType() {
        return foodType;
    }

    public void setFoodType(@NotNull FoodType f) {
        assert (foodType == null);
        assert (f.toString().equals(getTypedDataForColumn(FOOD_TYPE)));
        foodType = f;
    }

    @Override
    public Table<Food> getTable() {
        return Tables.FoodTable.instance();
    }

    public void addServing(@NotNull Serving s) {
        assert (getId().equals(s.getFoodId()));
        assert (equals(s.getFood()));
        assert (!servings.contains(s));
        servings.add(s);
    }

    @Nullable
    public Long getUsdaIndex() {
        return getTypedDataForColumn(USDA_INDEX);
    }

    @Nullable
    public String getNuttabIndex() {
        return getTypedDataForColumn(NUTTAB_INDEX);
    }

    @Nullable
    public Double getDensity() {
        return getTypedDataForColumn(DENSITY);
    }

    @Override
    public boolean equals(Object o) {
        return o instanceof Food && super.equals(o);
    }

    private String getDescriptionData(Column<Food, String> fieldName) {
        assert (DESCRIPTION_COLUMNS.contains(fieldName));
        return getTypedDataForColumn(fieldName);
    }

    public String getShortName() {
        return prettyFormat(false, false, false);
    }

    public String getLongName() {
        return prettyFormat(true, true, false);
    }

    public String getMediumName() {
        return prettyFormat(true, false, false);
    }

    public String getSortableName() {
        return prettyFormat(true, true, true);
    }

    public Long getCategoryId() {
        return getTypedDataForColumn(CATEGORY);
    }

    private String prettyFormat(boolean withBrand, boolean withDetails,
                                boolean sortable) {
        StringBuilder prettyName = new StringBuilder(getDescriptionData(NAME));

        String vp1 = getDescriptionData(VARIETY_PREFIX_1);
        String vp2 = getDescriptionData(VARIETY_PREFIX_2);
        String vs = getDescriptionData(VARIETY_SUFFIX);

        if (sortable) {
            if (hasDescriptionData(COMMERCIAL_NAME)) {
                prettyName.append(", ");
                if (withBrand && hasDescriptionData(BRAND)) {
                    prettyName.append(getDescriptionData(BRAND)).append(" ");
                }
                prettyName.append(getDescriptionData(COMMERCIAL_NAME));
            } else {
                if (hasDescriptionData(VARIETY_PREFIX_1) && hasDescriptionData(VARIETY_PREFIX_2)) {
                    prettyName.append(", ").append(vp1).append(" ").append(vp2);
                } else if (hasDescriptionData(VARIETY_PREFIX_1)) {
                    prettyName.append(", ").append(vp1);
                } else if (hasDescriptionData(VARIETY_PREFIX_2)) {
                    prettyName.append(", ").append(vp2);
                }
                if (hasDescriptionData(VARIETY_SUFFIX)) {
                    prettyName.append(", ").append(vs);
                }
            }
        } else {
            if (hasDescriptionData(COMMERCIAL_NAME)) {
                prettyName.insert(0, getDescriptionData(COMMERCIAL_NAME) + " ");
            } else {
                if (hasDescriptionData(VARIETY_PREFIX_2)) {
                    prettyName.insert(0, vp2 + " ");
                }
                if (hasDescriptionData(VARIETY_PREFIX_1)) {
                    prettyName.insert(0, vp1 + " ");
                }
                if (hasDescriptionData(VARIETY_SUFFIX)) {
                    prettyName.append(" ").append(vs);
                }
            }
            if (withBrand && hasDescriptionData(BRAND)) {
                prettyName.insert(0, getDescriptionData(BRAND) + " ");
            }
        }

        if (withDetails && hasDescriptionData(NOTES)) {
            prettyName.append(" (").append(getDescriptionData(NOTES)).append(")");
        }

        return prettyName.toString();
    }

    /*
     * Order of fields:
     * if sortable:
     *   if commercialName is present:
     *     <sqlName>, <brand> <commercialName> (<details>)
     *   else
     *     <sqlName>, <varPrefix1> <varPrefix2>, <varSuffix>, <brand> (<details>)
     * else:
     *   if commercialName is present:
     *      <brand> <commercialName> <sqlName> (<details>)
     *   else:
     *      <brand> <varPrefix1> <varPrefix2> <sqlName> <varSuffix> (<details>)
     */

    private Boolean hasDescriptionData(Column<Food, String> fieldName) {
        return hasData(fieldName);
    }

    @Nullable
    public Serving getDefaultServing() {
        Serving defaultServing = null;
        for (Serving s : servings) {
            if (s.isDefault()) {
                defaultServing = s;
            }
        }
        return defaultServing;
    }

    // quantity corresponds to that assumed by the data in the DB
    public NutritionData getNutritionData() {
        return nutritionData;
    }

    public void setNutritionData(@NotNull NutritionData nd) {
        assert (nutritionData == null);
        assert (getId().equals(nd.getFoodId()));
        nutritionData = nd;
    }

    public NutritionData getNutritionData(double quantity) {
        return nutritionData.rescale(quantity);
    }

    @Nullable
    public Serving getServingById(long servingId) {
        Serving serving = null;
        for (Serving s : servings) {
            if (servingId == s.getId()) {
                serving = s;
            }
        }
        return serving;
    }

    public List<Serving> getServings() {
        return new ArrayList<>(servings);
    }
}

File: src/main/kotlin/com/machfour/macros/entities/Food.kt
@file:Suppress("EqualsOrHashCode")

package com.machfour.macros.entities

import com.machfour.macros.core.*
import com.machfour.macros.entities.auxiliary.Factories
import com.machfour.macros.nutrients.FoodNutrientData
import com.machfour.macros.schema.FoodTable
import com.machfour.macros.schema.ServingTable
import com.machfour.macros.sql.RowData
import com.machfour.macros.sql.Table
import com.machfour.macros.units.GRAMS
import com.machfour.macros.units.MILLILITRES
import com.machfour.macros.units.UnitType

open class Food internal constructor(dataMap: RowData<Food>, objectSource: ObjectSource) :
        MacrosEntityImpl<Food>(dataMap, objectSource) {

    companion object {
        val descriptionColumns = listOf(
            FoodTable.NAME,
            FoodTable.BRAND,
            FoodTable.VARIETY,
            FoodTable.EXTRA_DESC,
            FoodTable.NOTES,
            FoodTable.INDEX_NAME,
            FoodTable.DATA_SOURCE,
            FoodTable.DATA_NOTES,
        )

        // Dynamically create either a Food or CompositeFood depending on the datamap passsed in.
        // Hooray for preferring static constructors over new!!!
        val factory: Factory<Food>
            get() = Factories.food
        
        val table: Table<Food>
            get() = FoodTable

        fun indexNamePrototype(
            basicName: String,
            brand: String?,
            variety: String?,
            extraDesc: String?,
        ): String {
            // use sortable name but replace sequences of spaces (and dashes) with a single dash
            return prettyFormat(
                basicName = basicName,
                brand = brand,
                variety = variety,
                extraDesc = extraDesc,
                withBrand = true,
                withVariety = true,
                withExtra = true,
                sortable = true
            )
                .replace(Regex("[()\\[\\]{}&%!$#@*^+=:;<>?/\\\\]"), replacement = "")
                .replace(Regex("[\\s-,]+"), replacement = "-")
                .removeSuffix("-")
        }

        /*
         * Order of fields:
         * if sortable:
         *     <name>, <brand>, <variety> (<extra desc>)
         * else:
         *     <brand> <variety> <name> (<extra desc>)
         */
        fun prettyFormat(
            basicName: String,
            brand: String?,
            variety: String?,
            extraDesc: String?,
            withBrand : Boolean = true,
            withVariety : Boolean = true,
            withExtra : Boolean = false,
            sortable : Boolean = false,
            includeEmptyFields: Boolean = true,
        ): String {
            val isPresent: (String?) -> Boolean = {
                it != null && (includeEmptyFields || it.isNotEmpty())
            }

            val prettyName = StringBuilder(basicName)
            if (sortable) {
                if (withBrand && isPresent(brand)) {
                    prettyName.append(", ").append(brand)
                }
                if (isPresent(variety)) {
                    prettyName.append(", ").append(variety)
                }
            } else {
                if (withVariety && isPresent(variety)) {
                    prettyName.insert(0, "$variety ")
                }
                if (withBrand && isPresent(brand)) {
                    prettyName.insert(0, "$brand ")
                }
            }
            if (withExtra && isPresent(extraDesc)) {
                prettyName.append(" (").append(extraDesc).append(")")
            }
            return prettyName.toString()
        }

        fun processFoodType(rowData: RowData<Food>): FoodType {
            val foodTypeString = rowData[FoodTable.FOOD_TYPE]
            requireNotNull(foodTypeString) { "Null food type string for rowData: $rowData" }
            val foodType = FoodType.fromString(foodTypeString)
            requireNotNull(foodType) { "Invalid food type string: $foodTypeString" }
            return foodType
        }
    }

    private val servingsInternal: MutableList<Serving> = ArrayList()

    val servings: List<Serving>
        get() = servingsInternal

    var defaultServing: Serving? = null
        private set

    // TODO check nutrient data is initialised
    open val nutrientData: FoodNutrientData = FoodNutrientData()

    val foodType: FoodType = processFoodType(dataMap)

    lateinit var foodCategory: FoodCategory
        private set

    open fun addNutrientValue(nv: FoodNutrientValue) {
        // TODO check ID matches
        nv.setFood(this)
        nutrientData[nv.nutrient] = nv
    }

    fun setFoodCategory(c: FoodCategory) {
        assert(!this::foodCategory.isInitialized && foreignKeyMatches(this, FoodTable.CATEGORY, c))
        foodCategory = c
    }

    override val table: Table<Food>
        get() = Companion.table

    override val factory: Factory<Food>
        get() = Companion.factory

    fun addServing(s: Serving) {
        assert(!servingsInternal.contains(s) && foreignKeyMatches(s, ServingTable.FOOD_ID, this))
        servingsInternal.add(s)
        if (s.isDefault) {
            setDefaultServing(s)
        }
    }

    private fun setDefaultServing(s: Serving) {
        assert(defaultServing == null) { "Default serving already set" }
        assert(servingsInternal.contains(s)) { "Serving does not belong to this food" }
        defaultServing = s
    }

    fun getServingById(id: Long): Serving? {
        return servingsInternal.firstOrNull { it.id == id }
    }

    fun getServingByName(name: String): Serving? {
        return servingsInternal.firstOrNull { it.name == name }
    }

    val naturalUnit: Unit
        get() = nutrientData.qtyUnit

    val validUnits: List<Unit>
        get() {
            return ArrayList<Unit>().apply {
                add(naturalUnit)
                // allow conversion if density is given
                if (density != null) {
                    if (naturalUnit.type === UnitType.VOLUME) {
                        add(GRAMS)
                    } else {
                        add(MILLILITRES)
                    }
                }
            }
        }

    // Returns list of quantity units and servings which can be used to measure this food
    val validMeasurements: List<PortionMeasurement>
        get() {
            return validUnits + servings
        }

    val usdaIndex: Int?
        get() = getData(FoodTable.USDA_INDEX)

    val nuttabIndex: String?
        get() = getData(FoodTable.NUTTAB_INDEX)

    val dataSource: String?
        get() = getData(FoodTable.DATA_SOURCE)

    val dataNotes: String?
        get() = getData(FoodTable.DATA_NOTES)

    val density: Double?
        get() = getData(FoodTable.DENSITY)



    override fun equals(other: Any?): Boolean {
        return other is Food && super.equals(other)
    }

    val shortName: String
        get() = prettyFormat(withBrand = false, withVariety = false)

    val longName: String
        get() = prettyFormat(withExtra = true)

    val mediumName: String
        get() = prettyFormat()

    val sortableName: String
        get() = prettyFormat(withExtra = true, sortable = true)

    val basicName: String
        get() = getData(FoodTable.NAME)!!

    val variety: String?
        get() = getData(FoodTable.VARIETY)

    val brand: String?
        get() = getData(FoodTable.BRAND)

    val extraDesc: String?
        get() = getData(FoodTable.EXTRA_DESC)

    val notes: String?
        get() = getData(FoodTable.NOTES)

    val indexName: String
        get() = getData(FoodTable.INDEX_NAME)!!

    val categoryName: String
        get() = getData(FoodTable.CATEGORY)!!

    val searchRelevance: Int
        get() = getData(FoodTable.SEARCH_RELEVANCE)!!

    private fun prettyFormat(
        withBrand : Boolean = true,
        withVariety : Boolean = true,
        withExtra : Boolean = false,
        sortable : Boolean = false
    ): String {
        return prettyFormat(
            basicName = basicName,
            brand = brand,
            variety = variety,
            extraDesc = extraDesc,
            withBrand = withBrand,
            withVariety = withVariety,
            withExtra = withExtra,
            sortable = sortable,

        )
    }

}